secure_tunneling/source/SecureTunnel.cpp

```diff
@@ -0,0 +1,140 @@
+            // Client callbacks
+            m_OnConnectionComplete = onConnectionComplete;
+            m_OnSendDataComplete = onSendDataComplete;
+            m_OnDataReceive = onDataReceive;
+            m_OnStreamStart = onStreamStart;
+            m_OnStreamReset = onStreamReset;
+            m_OnSessionReset = onSessionReset;
+
+            // Initialize aws_secure_tunneling_connection_config
+            aws_secure_tunneling_connection_config config;
+            AWS_ZERO_STRUCT(config);
+
+            config.allocator = allocator;
+            config.bootstrap = clientBootstrap ? clientBootstrap->GetUnderlyingHandle() : nullptr;
+            config.socket_options = socketOptions ? &socketOptions->GetImpl() : nullptr;
+
+            config.access_token = aws_byte_cursor_from_c_str(accessToken.c_str());
+            config.local_proxy_mode = localProxyMode;
+            config.endpoint_host = aws_byte_cursor_from_c_str(endpointHost.c_str());
+
+            config.on_connection_complete = s_OnConnectionComplete;
+            config.on_send_data_complete = s_OnSendDataComplete;
+            config.on_data_receive = s_OnDataReceive;
+            config.on_stream_start = s_OnStreamStart;
+            config.on_stream_reset = s_OnStreamReset;
+            config.on_session_reset = s_OnSessionReset;
+
+            config.user_data = this;
+
+            // Create the secure tunnel
+            m_secure_tunnel = aws_secure_tunnel_new(&config);
+        }
+
+        SecureTunnel::SecureTunnel(SecureTunnel &&from) noexcept
+        {
+            m_secure_tunnel = from.m_secure_tunnel;
+            from.m_secure_tunnel = nullptr;
+        }
+
+        SecureTunnel::~SecureTunnel()
+        {
+            if (m_secure_tunnel)
+            {
+                aws_secure_tunnel_release(m_secure_tunnel);
+            }
+        }
+
+        SecureTunnel &SecureTunnel::operator=(SecureTunnel &&rhs) noexcept
+        {
+            if (this != &rhs)
+            {
+                this->~SecureTunnel();
+
+                m_secure_tunnel = rhs.m_secure_tunnel;
+                rhs.m_secure_tunnel = nullptr;
+            }
+
+            return *this;
+        }
+
+        int SecureTunnel::Connect() { return aws_secure_tunnel_connect(m_secure_tunnel); }
+
+        int SecureTunnel::Close() { return aws_secure_tunnel_close(m_secure_tunnel); }
+
+        int SecureTunnel::SendData(const Crt::ByteCursor &data)
+        {
+            return aws_secure_tunnel_send_data(m_secure_tunnel, &data);
+        }
+
+        int SecureTunnel::SendStreamStart() { return aws_secure_tunnel_stream_start(m_secure_tunnel); }
+
+        int SecureTunnel::SendStreamReset() { return aws_secure_tunnel_stream_reset(m_secure_tunnel); }
+
+        aws_secure_tunnel *SecureTunnel::GetUnderlyingHandle() { return m_secure_tunnel; }
+
+        void SecureTunnel::s_OnConnectionComplete(void *user_data)
+        {
+            auto *secureTunnel = static_cast<SecureTunnel *>(user_data);
```

Comment:
reinterpret_cast is intended for pointer conversions. You're not intending for a possible `SecureTunnel* operator()(void *ptr)` to handle it.

Reply:
I will update these with a followup PR. Just want to merge this code for now to unblock UA development.